server/notifications.go

```diff
@@ -5,91 +5,87 @@
 package server
 
 import (
-	"context"
 	"sync"
-
-	"github.com/gofrs/uuid"
-)
-
-const (
-	// how many notifications can be in a notification buffer.
-	notificationBufferSize = 16
+	"sync/atomic"
 )
 
-type change struct {
-	// persistedIndex is a changed persisted index or `nil` if the value has not changed.
-	persistedIndex *int64
+// state is an extendable struct that stores
+// the information subscribers need to be notified about
+type state struct {
+	persistedIndex uint64
 }
 
-// Any returns true if there is a change.
-func (c change) Any() bool {
-	return c.persistedIndex != nil
+// Equals returns true if 2 instances of the state are equal
+func (s *state) Equals(v *state) bool {
+	return (s == nil && v == nil) || (s != nil && v != nil && s.persistedIndex == v.persistedIndex)
 }
 
 type notifications struct {
-	subscribers map[uuid.UUID]chan change
-	mutex       *sync.Mutex
+	mutex *sync.RWMutex
+	cond  *sync.Cond
+
+	// the whole state must be always replaced with `notify`
+	// the struct fields are not thread safe on their own
+	state *state
+
+	// the notifications are stopped when this flag > 0
+	stopped uint32
 }
 
-func (n *notifications) subscribe() (out <-chan change, stop func(), err error) {
-	ch := make(chan change, notificationBufferSize)
-	id, err := uuid.NewV4()
-	if err != nil {
-		return nil, nil, err
-	}
+// getState return the current state or `nil` if the initial state
+// is not set or if the notifications are stopped.
+// This is thread-safe.
+func (n *notifications) getState() *state {
+	n.mutex.RLock()
+	defer n.mutex.RUnlock()
+	return n.state
+}
 
-	n.mutex.Lock()
-	n.subscribers[id] = ch
-	n.mutex.Unlock()
+// wait returns a pointer to the current state or blocks until the
+// first state change or when notifications are stopped.
+func (n *notifications) wait() *state {
+	// before trying to lock we check the flag
+	// it should be less expensive
+	if atomic.LoadUint32(&n.stopped) != 0 {
```

Comment:
Have you tested that this is faster? The underlying mutex Lock() implementation already includes this optimization: https://cs.opensource.google/go/go/+/refs/tags/go1.18.4:src/sync/mutex.go;drc=de8101d21bcf5f1097bcfaf3a1b55820ba70dce9;l=81

```go
// Lock locks m.
// If the lock is already in use, the calling goroutine
// blocks until the mutex is available.
func (m *Mutex) Lock() {
	// Fast path: grab unlocked mutex.
	if atomic.CompareAndSwapInt32(&m.state, 0, mutexLocked) {
		if race.Enabled {
			race.Acquire(unsafe.Pointer(m))
		}
		return
	}
	// Slow path (outlined so that the fast path can be inlined)
	m.lockSlow()
}
```

Called from RWMutex.Lock() here: https://cs.opensource.google/go/go/+/refs/tags/go1.18.4:src/sync/rwmutex.go;l=139

Reply:
I don't follow, how would this optimisation affect the marked code? I'm checking whether the whole notifications thing is stopped before trying to lock a mutex. Why compete for the mutex if the notifications are stopped?

Reply:
My point was that mutexes are already implemented to avoid actually acquiring a lock unless actually necessary, so I was curious how much this is actually speeding up the implementation.

Are we saving 1 us in a process that completes normally in 1 ms? Or are we saving 1s in a process that normally takes 5s? If it is the first case I'm not sure the extra complexity is worth it.

It seems correct in this implementation though, so I don't feel that strongly about it.

Reply:
An atomic operation is always potentially faster than competing for a mutex, because acquiring a lock is not deterministic by its nature and it depends on the amount of go routines competing for the mutex. Theoretically it could block for seconds when an atomic operation is always the same cost.

This is what my comment is about.